pyrealm_build_data/synth_data.py

```diff
@@ -0,0 +1,81 @@
+"""This script uses a parametrized model to compress the input dataset.
+
+It fits a time series model to the input data and stores the model parameters.
+The dataset can then be reconstructed from the model parameters using the `reconstruct`
+function, provided with a custom time index.
+"""
+from typing import Tuple
+
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+VAR_BOUNDS = dict(
+    temp=(-25, 80),
+    patm=(3e4, 11e4),
+    vpd=(0, 1e4),
+    co2=(0, 1e3),
+    fapar=(0, 1),
+    ppfd=(0, 1e4),
+)
+
+
+def make_time_features(t: np.ndarray) -> pd.DataFrame:
+    """Make time features for a given time index."""
+    dt = pd.to_datetime(t).rename("time")
+    df = pd.DataFrame(index=dt).assign(const=1.0)
+
+    df["linear"] = (dt - pd.Timestamp("2000-01-01")) / pd.Timedelta("365.25d")
+
+    for f in [730.5, 365.25, 12, 6, 4, 3, 2, 1, 1 / 2, 1 / 3, 1 / 4, 1 / 6]:
+        df[f"freq_{f:.2f}_sin"] = np.sin(2 * np.pi * f * df["linear"])
+        df[f"freq_{f:.2f}_cos"] = np.cos(2 * np.pi * f * df["linear"])
+
+    return df
+
+
+def fit_ts_model(df: pd.DataFrame, fs: pd.DataFrame) -> Tuple[pd.DataFrame, float]:
+    """Fit a time series model to the data."""
+    df = df.dropna(axis=1, how="all").fillna(df.mean())
+    Y = df.values  # (times, locs)
+    X = fs.values  # (times, feats)
+    A = np.linalg.pinv(X) @ Y  # (feats, locs)
```

Comment:
`np.linalg.solve(X, Y)` is usually preferred? See for instance: https://stackoverflow.com/questions/31256252/why-does-numpy-linalg-solve-offer-more-precise-matrix-inversions-than-numpy-li

Reply:
`np.linalg.solve` only accepts square matrices. The corresponding function for non-square matrices is `np.linalg.lstsq`. I tried it in place of `np.linalg.pinv`, and it only saves ~0.06s for a single variable. However, I've used it in the code instead now.